Validation for revised expressions.

kennethtran93 opened the pull request:
This adds validation for editing existing expressions on the Expression Table.  Resolves #849.

Note:  This is only when editing expressions, NOT applied (yet) to adding new expressions.

This only covers the more common items such as:
- Validating Regex
- Checking for Start AND End Slash for Regex
- If not Regex:
  - slash is found at beginning or end.
  - comma is found (only allowed in Regex)

If there are other validations I should do please let me know.

This also moves the reset/save buttons to the bottom of the input field and makes them larger.

![image](https://user-images.githubusercontent.com/6724477/97068099-51a62080-1579-11eb-86ad-87ef1243c6b5.png)

![image](https://user-images.githubusercontent.com/6724477/97068175-3b4c9480-157a-11eb-8bfc-8d38710769b6.png)

Since the first commit, the following items have been added/adjusted:
- Wrapping Long expression domains through textarea to almost mimic the size of the input field.  Resolves #850.
- Shift+Enter to add to GreyList on active list.
- Validate New Expressions from manual input as well as from file import.
- Fixed crash when having a `\` at the end of a search string.

![Peek 2020-10-26 22-45](https://user-images.githubusercontent.com/6724477/97264216-c2e80c80-17e1-11eb-9abf-efe75e6b8e4b.gif)


Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Validation for revised expressions.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Fix typo

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

ke-d commented on src/ui/common_components/ExpressionTable.tsx:
```diff
@@ -88,6 +99,45 @@ class ExpressionTable extends React.Component<
+    if (!inputTrim) {
+      return this.setInvalid(browser.i18n.getMessage('inputErrorEmpty'));
+    }
+    if (inputTrim.startsWith('/') && inputTrim.endsWith('/')) {
```
Hmm this seems like something you should probably unit test to verify the results.

kennethtran93 replied:
Moved to a function in LIbs to allow validation from List of Expressions (new input) as well.

kennethtran93 committed:
Change non-edit text to disabled input for wrapping long values.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Revert "Fix typo"

This reverts commit cf2333ee

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Change non-edit text to readonly textarea for wrapping long values.  This should make the overall width from either state to be very similar.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Move Domain Expression Validation to Libs to allow other future uses.  Also prevent validation errors from default browser validation.  This adds one more check for any spaces in between words.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
This adds the ability to use Shift+Enter to add expressions to GREYLIST from the list of Expressions.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Validate New Expressions.
Show errors for each one in a new line.
Fix typo in validation function.
Add css necessary to allow line breaks '\n' in alerts.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Apply Validation from imported files.
Add success and error status message to imported and manually added expressions.
Make 'File Not Found' translatable.
Added Import Expression File Checks.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Minor search enhancement.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Prevent a crash when searching expressions with '\' at the end.
Minor status enhancement.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Prevent additional crashes when searching for expressions.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>
